File: DiveIntoDeepLearning/Chapter07/MySequential.py
from torch import nn


class MySequential(nn.Module):
    def __init__(self, *args):
        super().__init__()

        for idx, module in enumerate(args):
            self.add_module(str(idx), module)

    def forward(self, X):
        for module in self.children():
            X = module(X)

        return X

File: DiveIntoDeepLearning/Chapter07/TorchSharpChapter08/MySequential.cs
using TorchSharp;
using static TorchSharp.torch;
using static TorchSharp.torch.nn;

namespace TorchSharpChapter08;

public class MySequential : Module<Tensor, Tensor>
{
    public MySequential(params Module[] args) : base(nameof(MySequential))
    {
        foreach ((Module layer, string i) in args.Select((x, i) => (x, i.ToString())))
        {
            add_module(i.ToString(), layer);
        }
    }

    public override Tensor forward(Tensor input)
    {
        foreach (Module module in children())
        {
            if (module is not Module<Tensor, Tensor> layer)
            {
                throw new InvalidOperationException("Modules must all accept and return tensors");
            }

            input = layer.call(input);
        }

        return input;
    }
}